Highlight destination identifier on in-page navigation. #310

nickrobinson opened the pull request:
Adding 2 seconds highlight for the element that was clicked in the index list.

nickrobinson commented:
I signed it!

nickrobinson commented:
I signed it!

wallclockbuilder commented on gddo-server/assets/site.js:
```diff
@@ -119,6 +119,35 @@ $(function() {
+        return s.replace( /(:|\.|\[|\])/g, "\\$1" );
+    }
+
+    $( ".indexEntry").click(function() {
```
Inconsistent space after opening parens.

wallclockbuilder commented:
As discussed in the issue #310
This is a straight-forward fix that should highlight the target identifier on navigation.
Kindly remove the timed fadeout. That defeats the purpose.

shurcooL commented:
If the highlight is going to be static (instead of fading out), please add a Escape key shortcut to clear it. It can be distracting and I'm used to be able to press Escape to clear such things.

As an example of what I mean, see http://gotools.org/github.com/golang/gddo/gddo-server#background-go-L1-L21. Pressing Escape clears the highlighted lines.

margaery commented:
The highlight is distracting and escape is not discoverable. If automatic fade is not acceptable, how about fade on scroll?

nickrobinson commented:
I have updated the pull-request to add an escape key even to remove the highlight. My fear on the fade by scrolling is that the initial problem would remain unsolved. If scrolling causes the highlight to go away the highlight may be gone by the time you get to the entry you are looking for.

margaery commented:
My idea is to start the fade on scroll after the jump is complete. With this approach, the item is highlighted at least as long as in the original PR.  

Another option is fade automatically, but with a longer fade time.

The keyboard shortcuts should also highlight the item.

wallclockbuilder commented:
Thanks for the quick turn around @nickrobinson 

Tested this. Doesn't work as expected.
To reproduce:
1) Select an item from the index. (Page navigates to target, with the title highlighted.)
2) Go back to the index list and select a second item. (Page navigates to second target and highlights it. Now there are two highlighted targets on page.)
Find a screenshot below:

![highlight_malfunction](https://cloud.githubusercontent.com/assets/2033945/10125546/2537d672-6568-11e5-8d52-dafa1ceb9ee2.png)


I suggest you fix this by only highlighting the selected target. At any given time there should only be one selected target on page.

nickrobinson commented:
@wallclockbuilder, Updated to address the issue of the highlight hanging around when clicking a new element. I am wondering if the same highlighting should be done for the permalinks as well.

<img width="867" alt="screen shot 2015-09-27 at 7 59 20 pm" src="https://cloud.githubusercontent.com/assets/479287/10125888/51420e7e-6552-11e5-8284-3d4a8e839ca9.png">
<img width="852" alt="screen shot 2015-09-27 at 7 59 34 pm" src="https://cloud.githubusercontent.com/assets/479287/10125889/5144ce98-6552-11e5-9984-9bb5a3c96775.png">

adg commented:
I have a counterproposal: make the highlighting a bit less obnoxious (pale yellow background?), don't fade it out, and don't bother providing a mechanism to clear it.

margaery commented:
If the highlight does not fade out, then the [onhashchange](https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onhashchange) event is a good place to add & remove highlight. This will catch navigation through the keyboard interface in addition to clicks on the elements hooked in this PR.

nickrobinson commented:
Update the PR. I moved the highlighting CSS to the site.css file and changed the color to #FDFF9E. I now update the class for the element when either the permalink or index link is clicked. Moved the highlighting functionality to a separate function to reduce redundancy.

margaery commented:
The keyboard interface should also highlight the target.  The hashchange event will catch everything.

    var highlightedSel = "";
    if ("onhashchange" in window) {
        window.onhashchange = function() {
            if (highlightedSel) {
                $(highlightedSel).removeClass("highlighted");
            }
            highlightedSel = window.location.hash.replace( /(:|\.|\[|\]|,)/g, "\\$1" );
            if (highlightedSel) {
                $(highlightedSel).addClass("highlighted");
             }
        }
    }

adg commented on gddo-server/assets/site.js:
```diff
@@ -119,6 +119,29 @@ $(function() {
+        return s.replace(/(:|\.|\[|\])/g, "\\$1");
+    }
+
+    highlightElement = function(el) {
```
```
var hl = function(el) {
 ..
}
```

adg replied:
```
$(".indexEntry, .permalink").click(function(el) {
 ...
})
```

shurcooL commented:
> and don't bother providing a mechanism to clear it.

I am afraid there's a very high chance that will bug me... :( I can reserve my judgement and wait and see, but I'm pretty sure I will find it distracting and be upset if I can't just hit Escape to clear it.

adg commented:
@shurcooL well let's see how it goes.

wallclockbuilder commented:
Perhaps, pale yellow like the github line highlight? ![screen shot 2015-09-29 at 5 11 20 am](https://cloud.githubusercontent.com/assets/2033945/10155761/93d5f1d0-6668-11e5-8434-b7fb7729f717.png)

shurcooL commented:
> If the highlight does not fade out, then the onhashchange event is a good place to add & remove highlight. This will catch navigation through the keyboard interface in addition to clicks on the elements hooked in this PR.

Yes, the onhashchange event is the right place to do this. You'll likely also want to do it on DOMContentLoaded event or something similar in order to do it when a page is freshly loaded (from an external link).

margaery commented:
@shurcooL I updated the code above to execute after load. Thanks.

nickrobinson commented:
Updated to add @margaery hashchange code.

adg commented on gddo-server/assets/site.js:
```diff
@@ -119,6 +119,40 @@ $(function() {
+    }
+
+    $(".indexEntry", ".permalink").click(function() {
+        highlightElement(this);
```
why are you calling the highlightElement function instead of just putting the body of the function here?

margaery replied:
The click handler can be removed.  The hashchange event fires when these elements are clicked.

nickrobinson commented:
ClickHandler removed and highlightElement function removed

garyburd commented:
I pushed a demo of the PR to http://gary.burd.info/demo/.  

The PR [highlights entire examples](http://gary.burd.info/demo/#example-StripPrefix).  The highlighting should either be restricted to the example header or examples should not be highlighted at all. 

Does it make sense to limit highlighting to exported identifiers?  

    ...
    if (/^[^_][^-]*$/.test(iwindow.location.hash)) {
        highlightedSel = window.location.hash.replace( /(:|\.|\[|\]|,)/g, "\\$1" );
    } else {
        highlightedSel = ""
    }
    ...

The test regex in this snippet is copied from [the find modal](https://github.com/golang/gddo/blob/8dab6303bc8c4444aa8e8106bb3acad76c4450b7/gddo-server/assets/site.js#L73).

nickrobinson commented:
I think it makes more sense to not highlight the examples at all. I removed the class entry from the example field.

adg commented:
The change looks good to me. Can you please squash your commits into one?

garyburd commented:
Examples are still highlighted with the most recent change.

The classes added to gddo-server/assets/templates/pkg.html are not used. This file can be removed from the PR.

nickrobinson commented:
Re-based and added change to ignore example entries. It currently looks for a match with the name "example-"

adg commented:
Thanks

shurcooL commented:
When will this change be deployed? I'd like to try it out. :)

wallclockbuilder commented:
Looks good to me! Great work @nickrobinson 

![screen shot 2015-10-01 at 11 46 25 am](https://cloud.githubusercontent.com/assets/2033945/10219858/1620febc-6832-11e5-9a89-1a299a523048.png)

shurcooL commented:
> > and don't bother providing a mechanism to clear it.
>
> I am afraid there's a very high chance that will bug me... :( I can reserve my judgement and wait and see, if you wish, but I'm pretty sure I will find it distracting in some situations and be upset if I can't just hit Escape to clear it.

I wanted to follow up and admit that I am quite happy with the current behavior and don't feel a strong need to be able to press Escape to clear the selection. Primarily because it's quite unobtrusive.

Great work on this feature indeed!
